prusti-viper/src/encoder/snapshot_encoder.rs

```diff
@@ -630,10 +658,56 @@ impl<'p, 'v, 'r: 'v, 'a: 'r, 'tcx: 'a> SnapshotEncoder<'p, 'v, 'tcx> {
     fn encode_snap_arg(&self, location: vir::Expr, field: vir::Field, field_ty: ty::Ty<'tcx>)
         -> EncodingResult<vir::Expr>
     {
-        let snapshot = self.encoder.encode_snapshot(field_ty)?;
-        Ok(snapshot.snap_call(
-            self.encode_arg_field(location, field)
-        ))
+        warn!("encode_snap_arg begin");
+        let res = match field_ty.kind() {
+        ty::TyKind::Adt(adt_def, _) if adt_def.is_box() => { //TODO this breaks for lots of boxes. e.g. Box<Box<T>> or Box<i32>
+            let boxed_ty = field_ty.boxed_ty();
+            warn!("enocding boxed ty {:?}", boxed_ty);
```

Comment:
Can you invoke the snapshot encoding recursively here, i.e. `self.encoder.encode_snapshot(boxed_ty)`?

Reply:
I think that would cause infinite recursion. For example if I have a struct
```rust
struct List {
    value: u32,
    next: Option<Box<List>>,
}
```
`encode_snapshot(List)` would call `encode_snapshot(Option<Box<List>>)` and then that would call `encode_snapshot(Box<List>)` which calls `encode_snap_arg(Box<List>)` and then `encode_snapshot(List)` (because we'd call `self.encoder.encode_snapshot(boxed_ty)` where the `boxed_ty` is `List` )

This admittedly rather ugly code I wrote is an attempt to prevent this infinite recursion by only computing the parts of the snapshot we actually need.

Reply:
I understand the issue, but I just doing another (less complete) copy of the encoding for boxes is not really a solution. You could split ADT encoding into two phases, such that the `snapshot` map in `Encoder` contains an entry for the snapshot being encoded before the recursive case is hit.